components/daikin_s21/climate/daikin_s21_climate.cpp

```diff
@@ -192,161 +249,153 @@ void DaikinS21Climate::set_supported_modes_override(std::set<climate::ClimateMod
-  if (!this->use_room_sensor()) {
-    return 0.0F;
+/**
+ * Determine the S21 setpoint value based on the current temperature and target temperature.
+ *
+ * Applies an offset from the external reference sensor if present.
+ */
+DaikinC10 DaikinS21Climate::calc_s21_setpoint() {
+  // First ensure we're in a setpoint mode
+  if (is_setpoint_mode(this->mode) == false) {
+    return TEMPERATURE_INVALID;
   }
-  float room_val = this->room_sensor_degc();
-  float s21_val = this->get_parent()->get_temp_inside();
-  return s21_val - room_val;
-}
-
-float nearest_step(float temp) {
-  return std::round(temp / SETPOINT_STEP) * SETPOINT_STEP;
-}
-
-// What setpoint should be sent to s21, acconting for external room sensor.
-float DaikinS21Climate::calc_s21_setpoint() {
-  float offset_target = this->target_temperature + this->get_room_temp_offset();
-  return nearest_step(offset_target);
-}
-
-// How far from desired setpoint is the current S21 setpoint?
-float DaikinS21Climate::s21_setpoint_variance() {
-  return abs(this->get_parent()->get_setpoint() - this->calc_s21_setpoint());
-}
 
-void DaikinS21Climate::save_setpoint(float value, ESPPreferenceObject &pref) {
-  int16_t stored_val = static_cast<int16_t>(value * 10.0F);
-  pref.save(&stored_val);
+  const auto current_temperature = this->get_current_temperature();
+
+  // offset the ideal setpoint by the difference between the unit and reference sensor (0 if the same sensor)
+  auto s21_setpoint = static_cast<DaikinC10>(this->target_temperature) + (this->get_parent()->get_temp_inside() - current_temperature);
+
+  // Round to Daikin's internal setpoint resolution
+  // When the ideal setpoint is between steps force it in the direction of change by controlling rounding. Over time it should oscillate over the ideal setpoint.
+  if (s21_setpoint < current_temperature) {
+    // commanding the unit lower, round down
+  } else if (s21_setpoint > current_temperature) {
+    // commanding the unit higher, round up by adding almost a full step
+    s21_setpoint = s21_setpoint + (SETPOINT_STEP - 1);
+  } else {
+    // no difference, round to nearest
+    s21_setpoint = ((s21_setpoint * 2) + SETPOINT_STEP) / 2;  //add a half step (scaled for full precision)
+  }
+  s21_setpoint = (s21_setpoint / SETPOINT_STEP) * SETPOINT_STEP;  // complete round by truncating fractional component of step
+
+  // Ensure it's valid for the unit's current mode.
+  // Daikin will clamp internally with a slightly out of range value, but it's faster for the UI to do it here without waiting for comms
+  // Also, when a large external temperature offset is used, the value can be so far out of range it will be NAK'd
+  return std::clamp(s21_setpoint,
+      (this->mode == climate::CLIMATE_MODE_HEAT) ? this->min_temperature : this->min_cool_temperature,
+      (this->mode == climate::CLIMATE_MODE_COOL) ? this->max_temperature : this->max_heat_temperature);
 }
 
-void DaikinS21Climate::save_setpoint(float value) {
```

Comment:
Simplify setpoint handlers, just store and restore the int16_t value directly. Use NAN sentinel value instead of std::optional.